File: spider/chinalaw/go/ocr/baidu.go
package ocr

import (
	"os"
	"encoding/json"
	//"encoding/base64"
    "fmt"
	"net/http"
	"net/url"
	"io/ioutil"
	"strings"
	. "../bs64"
)
//https://ai.baidu.com/ai-doc/OCR/1k3h7y3db
//
var(
	BAIDU_TK_SERVER="https://aip.baidubce.com/oauth/2.0/token"
	BAIDU_OCR_SERVER="https://aip.baidubce.com/rest/2.0/ocr/v1/accurate_basic"
	client_id="1111"
	client_secret="2222"
)

type Token struct {
	RefreshToken  string `json:"refresh_token"` 
	ExpiresIn     int64  `json:"expires_in"`    
	Scope         string `json:"scope"`         
	SessionKey    string `json:"session_key"`   
	AccessToken   string `json:"access_token"`  
	SessionSecret string `json:"session_secret"`
}

type OcrResult struct {
	LogID          int64         `json:"log_id"`          
	Direction      int64         `json:"direction"`       
	WordsResultNum int64         `json:"words_result_num"`
	WordsResult    []WordsResult `json:"words_result"`    
}

type OcrFail struct {
	ErrorCode int64  `json:"error_code"`
	ErrorMsg  string `json:"error_msg"`
}

type WordsResult struct {
	Words       string      `json:"words"`      
	Probability Probability `json:"probability"`
}

type Probability struct {
	Variance int64 `json:"variance"`
	Average  int64 `json:"average"` 
	Min      int64 `json:"min"`     
}


func Env(a string, b string) (string){
	s:=os.Getenv(a) 
	if len(s)>0 {
		return s
	}else{
		return b
	}
}


func qs(u string,o map[string]string) string{
	 Url, _ := url.Parse(u)
	 p:=url.Values{}
	 for k, v := range o{
	 	p.Add(k,v)
	 }
	 Url.RawQuery = p.Encode()
	 return Url.String()
}

func qs1(o map[string]string) string{
	 p:=url.Values{}
	 for k, v := range o{
	 	p.Add(k,v)
	 }
	 return p.Encode()
}


func GetToken() (Token,error){
     var r Token
	 u:=BAIDU_TK_SERVER
	 o:=map[string]string{
	   "grant_type": "client_credentials",
	   "client_id": Env("baidu_ocr_APIKey",client_id),
	   "client_secret":Env("baidu_ocr_SecretKey", client_secret),
	 }
	 u1:=qs(u,o)

     resp,err := http.Get(u1)
	 if err!=nil{
		 return r,err
	 }
     defer resp.Body.Close()
     body, _ := ioutil.ReadAll(resp.Body)

	 err = json.Unmarshal(body, &r)

	 //fmt.Println("[token] ",r)
	 return r, err
	
}


func Ocr(tk string , s string) (OcrResult,error){
	u:=BAIDU_OCR_SERVER
	o:=map[string]string{
		"access_token":tk,
		"image":s,
	}
	//fmt.Println("[oo] ",o) 
	q:=qs1(o)
	req, err := http.NewRequest("POST", u,strings.NewReader(q))
	if err!=nil{
		fmt.Println("[网不行] ",err) 
	}
    client := http.Client{}
    response, e := client.Do(req)
	if e != nil {
		fmt.Println("[被封了] ",e) 
	}
	defer response.Body.Close()
	b, _ := ioutil.ReadAll(response.Body)
	fmt.Println("[ocr] ",string(b)) 
	var r OcrResult
	err = json.Unmarshal(b, &r)
	return r, err
}


func Reco() string{
	file:="code.jpg"
	s:=Read_base64(file)
	tk,_:=GetToken()
	r,e:=Ocr(tk.AccessToken,s)
	if e!=nil {
		fmt.Println("[超了。。] ",e) 
        panic(e)
	}
	fmt.Println(r) 
	w:=""
	if len(r.WordsResult)>0 {
		w1:=r.WordsResult[0].Words
		w2:=strings.Replace(w1, " ", "", -1)
		w3:=strings.Replace(w2, "-", "", -1)
		if len(w3) == 5{
			w=w3
		}else{
			fmt.Println("[不对]",w3) 
		}
	}
	fmt.Println("oooo",w) 
	return w
}

File: spider/chinalaw/rust/ocr/src/baidu.rs
extern crate serde_json;
use std::process;
use std::env;
use std::fs;
use std::fs::File;
use std::error::Error;
use std::io::prelude::*;
use std::path::Path;
use std::io;
use std::io::prelude::*;
use std::io::BufReader;

use std::string::*;
use std::collections::*;
use std::collections::HashMap;
//use std::io::{stdout, Write};
//use std::io::Read;

use bytes::Bytes;
//extern crate url;
//use url::form_urlencoded;
use std::path::PathBuf;

use async_std::task;
use futures::future::BoxFuture;
use surf::middleware::{HttpClient, Middleware, Next, Request, Response};
use surf::mime;
//use serde::{Deserialize, Serialize};
use serde_derive::{Serialize, Deserialize};
use headers::{Cookie, HeaderMap, HeaderMapExt, HeaderValue};




#[derive(Deserialize, Serialize)]
pub struct Config {
    grant_type: String,
    client_id: String,
    client_secret: String,
}

#[derive(Serialize, Deserialize)]
pub struct Token {
    #[serde(rename = "refresh_token")]
    refresh_token: String,

    #[serde(rename = "expires_in")]
    expires_in: i64,

    #[serde(rename = "scope")]
    scope: String,

    #[serde(rename = "session_key")]
    session_key: String,

    #[serde(rename = "access_token")]
    access_token: String,

    #[serde(rename = "session_secret")]
    session_secret: String,
}

#[derive(Serialize, Deserialize)]
pub struct TokenFail {
    #[serde(rename = "error_code")]
    error_code: i64,

    #[serde(rename = "error_msg")]
    error_msg: String,
}

//{"error_code":17,"error_msg":"Open api daily request limit reached"}
#[derive(Serialize, Deserialize)]
pub struct RecoResultFail {
    #[serde(rename = "error_code")]
    error_code: i64,

    #[serde(rename = "error_msg")]
    error_msg: String,
}


#[derive(Serialize, Deserialize)]
pub struct RecoResult {
    #[serde(rename = "log_id")]
    log_id: i64,

    #[serde(rename = "direction")]
    direction: i64,

    #[serde(rename = "words_result_num")]
    words_result_num: i64,

    #[serde(rename = "words_result")]
    words_result: Vec<WordsResult>,
}

#[derive(Serialize, Deserialize)]
pub struct WordsResult {
    #[serde(rename = "words")]
    words: String,

    #[serde(rename = "probability")]
    probability: Probability,
}

#[derive(Serialize, Deserialize)]
pub struct Probability {
    #[serde(rename = "variance")]
    variance: i64,

    #[serde(rename = "average")]
    average: i64,

    #[serde(rename = "min")]
    min: i64,
}

lazy_static! {

    static ref BAIDU_TK_SERVER:String="https://aip.baidubce.com/oauth/2.0/token".to_owned();
    static ref BAIDU_OCR_SERVER:String="https://aip.baidubce.com/rest/2.0/ocr/v1/accurate_basic".to_owned();
 //   static ref CONFIG: Config= {
 //       let client_id=env::var("baidu_ocr_APIKey").unwrap();
 //       let client_secret=env::var("baidu_ocr_SecretKey").unwrap();
 //       let query= Config { 
 //           grant_type:"client_credentials".into(),
 //           client_id: client_id.into(),
 //           client_secret:client_secret.into(),
 //       };
 //       query
 //   };
}


pub fn get_config()->Config{
        let client_id=env::var("baidu_ocr_APIKey").unwrap();
        let client_secret=env::var("baidu_ocr_SecretKey").unwrap();
        let query=Config { 
            grant_type:"client_credentials".into(),
            client_id: client_id.into(),
            client_secret:client_secret.into(),
        };
        query
}


pub fn read_base64(file_name:&str) -> String{
   //println!("cccccccc{}" , env::current_dir().unwrap().display());
   //println!("curexe: {}" , env::current_exe().unwrap().display());
   let t=env::current_dir().unwrap().display().to_string();
   let p: PathBuf = [&t, file_name].iter().collect();
   let p1=p.to_str().unwrap();
   println!("ccccccccc{}" ,p1);
   let mut f = File::open(p1).unwrap();
   let mut buffer = Vec::new();
   f.read_to_end(&mut buffer);
   let c=Bytes::from(buffer);
   let d=base64::encode(&c);
   d
}

pub async fn get_token() -> Result<String, Box<dyn std::error::Error + Send + Sync + 'static>>{
    let u=&*BAIDU_TK_SERVER;
    let Token {
        access_token,
refresh_token, expires_in, scope, session_key, session_secret
    } = surf::get(u)
            .set_query(&get_config())?
            .recv_json()
            .await?;
     println!("{}", access_token);
     Ok(access_token)
}

pub async fn reco(file_name:&str) -> Result<String, Box<dyn std::error::Error + Send + Sync + 'static>>{
    #[derive(Deserialize, Serialize)]
    struct Img {
        access_token: String,
	image:String,
    }
    let s=read_base64(file_name);

    let tk=get_token().await?;
    let query= &Img { 
        access_token: tk.into(),
	image:s.into(),
    };
    let u=&*BAIDU_OCR_SERVER;
    let res=surf::post(u)
         .body_form(&query)?
         .recv_string()
         .await?;
     Ok(res) 
}




